atlite/wind.py

```
@@ -115,14 +123,172 @@ def extrapolate_wind_speed(
+                break
+        else:
+            raise AssertionError(
+                "None of the datasets modules define retrieve_windspeed_average"
+            )
+
+        logger.info(
+            "Retrieving average windspeeds at %d from module %s", height, module
+        )
+        data_average = retrieve_windspeed_average(cutout, height)
+
+    if isinstance(real_average, str | Path):
+        real_average = rio.open(real_average)
+
+    if isinstance(real_average, rio.DatasetReader):
+        real_average = rio.band(real_average, 1)
+
+    if isinstance(real_average, rio.Band):
+        transform = _as_transform(data_average.indexes["x"], data_average.indexes["y"])
+
+        real_average, _ = rio.warp.reproject(
+            real_average,
+            np.empty(data_average.shape),
+            dst_crs=CRS(data_crs),
+            dst_transform=transform,
+            dst_nodata=np.nan,
+            resampling=rio.enums.Resampling.average,
+        )
+
+        real_average = xr.DataArray(real_average, data_average.coords)
+    else:
+        raise ValueError(
+            f"`real_average` needs to be a path or rasterio object to a raster dataset"
+            f", but is: {real_average}"
+        )
+
+    return (real_average / data_average).assign_attrs(height=height)
+
+
+def apply_windspeed_bias_correction(
+    ds: xr.Dataset, windspeed_bias_correction: bool | xr.DataArray | None = None
+) -> tuple[xr.Dataset, int | None]:
+    """
+    Apply bias correction to windspeeds.
+
+    uses either an explicitly provided scaling factor or a pre-calculated
+    scaling factor supplied in cutout.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset containing wind speed time-series (at one or multiple heights)
+        with keys like 'wnd{height:d}m' and potentially a pre-calculated scaling
+        factor at key 'wnd_bias_correction'
+    windspeed_bias_correction : bool or DataArray, optional
```

I am also not a fan of these "could be anything" arguments. 
`False` is not needed, you just don't execute the function.
`None` uses a default from ds if it exists, otherwise does nothing is not clean either. This logic can be done before. You either use it or you don't by not calling this method. `None` can be used, but if you rely on a default it should raise an Error if the data for that does not exist. Same goes for `True`.

---

The default has been chosen to not introduce any problems for existing uses.

---

But 2-4 can be handled before calling all those `apply_`... methods that may do nothing. 
It just makes it much harder to follow the code. Mypy will also be a pain to implement and less helpful with these many typed variables. I am not saying that this needs to be changed here, but in an ideal world we would have most of the variables in a single type. For sure internally

---

Ok, moved the argument treatment up into the cutout.wind method. Does this make your mypy life less painful?